### ISSUE TITLE
Language Service: TypeParameter.constraint is lazily calculated, but does not have an accessor function

### ISSUE THREAD

**Issue by fsoikin** [microsoft/TypeScript#5689]
Because of this, I have to go through this awkward motion every time:

``` typescript
function getConstraint( type: ts.TypeParameter ) {
  // This call will cause the typechecker to resolve properties, as well as a bunch of other information
  // about the type (such as generic constraints), but we don't actually need its result right now.
  type.getProperties(); 

  return type.constraint;
}
```

For some members (such as "properties" - above), there is an accessor function, which I can use to be sure the data is up to date. Another category of members don't need an accessor function, because they are calculated eagerly, and so are "always there".

But there is this third category of members, which fall through the gap. One example is `TypeParameter.constraint`. Another is `InterfaceTypeWithDeclaredMembers.declaredProperties`. There are probably more, but I haven't come across them yet.

**Comment by mhegazy** [comment]
Would exposing [`getConstraintOfTypeParameter`](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L4083-L4094) on the API address the issue?

**Comment by fsoikin** [comment]
It would solve half the issue: now I can use that function to get the value while making sure it's calculated. But that is assuming that _I know which function to use_. If I was just exploring the API, I would see the `constraint` property and wouldn't think twice about just using it. It will be only after hours of debugging and googling (and possibly finding this issue) that I would gain the necessary knowledge.

So to solve the other half, the property needs to be hidden. But I'm not sure if it will break anything somewhere.

**Comment by mhegazy** [comment]
Agreed. so what we need to do is 1. expose the getConstraintOfTypeParameter on the TypeChecker interface, 2, add Type.getConstraint() method on the TypeObject class, and 3, mark type.constraint as internal.

3 is a breaking change, so not sure if/when we can do it. but other ones should be doable.

**Comment by mhegazy** [comment]
PRs are definitely welcomed.

**Comment by RyanCavanaugh** [comment]
This was fixed via #20137
